[bug 1146982] Add 'Send to Device' macro

alexgibson opened the pull request:
This is still a work in progress, do DNM for now.

alexgibson commented:
@pmclanahan - I'm as far along with this design/template wise as I think I can get for now. We should be good to start adding in the back end logic whenever you're ready. Feel free to make any changes you need, or ping me if you have any questions.

flodolo commented on bedrock/base/templates/macros.html:
```diff
@@ -642,3 +642,83 @@
+            </div>
+          </div>
+          <p class="legal sms">
+            {{ _('SMS service available to US phone numbers only. SMS &amp; data rates may apply.')}} {{ _('The intended recipient of the email or SMS must have consented.')}} <a href="{{ url('privacy.notices.websites') }}#campaigns" class="more">{{ _('Learn more')}}</a>
```
US or U.S. (line 661)?

alexgibson replied:
Good catch, probably best to go for U.S. but will double-check.

alexgibson replied:
We already use U.S elsewhere in the form, so will use that here

pmclanahan commented:
Thanks Alex. I'll get started. Do you have a good handle on what exactly is required? The bug is a bit confusing to me and it seems like maybe you've had meetings. If you know more would you mind summarizing here or in the bug?

Seems like we need an AJAX endpoint for the form submission since this will be placed on several pages, and will need to accept email or phone numbers. Anything other than that?

Also, I noticed that you hard coded the list of locales, but I think we can generate that. There's already a method in `bedrock.newsletter.utils` that will give us the list of newsletter languages, and we can find all of the locales using that if we want it to be a bit more future proof. I'll do that if it's something you think would be good.

alexgibson commented:
> Seems like we need an AJAX endpoint for the form submission since this will be placed on several pages, and will need to accept email or phone numbers. Anything other than that?

Yes, basically the form needs to accept a single input, which can be either an email address of a telephone number. Only folks in the US get the SMS messaging.

There are also a few different ways the form can be used to generate different emails. Depending on the page where the form is used, it can either:

* Send an Email or SMS that links to both iOS and Android app stores in the same message.
* Send an Email or SMS linking to just the iOS version on the App Store.
* Send an Email or SMS linking to just the Android version on Google Play.
* Let's the user select their chosen OS before submitting, via a drop down menu in the form.

I've added some docs on how the widget can be used in the PR, which goes into more detail.

If you're unsure about anything or have any more questions, I'm happy to try and help.

> Also, I noticed that you hard coded the list of locales, but I think we can generate that. There's already a method in bedrock.newsletter.utils that will give us the list of newsletter languages, and we can find all of the locales using that if we want it to be a bit more future proof. I'll do that if it's something you think would be good.

The list of locales is what I've been provided with. AFAIK, these are the page locales that should see this widget for the spring campaign release. I don't know if that will change, but if you would prefer to use the above method that's fine with me.

ckprice commented:
@pmclanahan what are the locales in `bedrock.newsletter.utils`?

pmclanahan commented:
@ckprice they come from basket. bedrock loads https://basket.mozilla.org/news/newsletters/ and creates a unique list of all languages listed for newsletters. So using that means that we'll keep up to date with what the newsletters actually support. It's the same place that the newsletter forms get their language drop down lists.

flodolo commented:
One thing that I'm missing is how to manage this with lang files. It doesn't make sense to have these string in a shared file, since we need them only for a bunch of locales. Would it work with a ```/base/templates/macros.lang``` file, or do we need a new shared file in ```DOTLANG_FILES```?

pmclanahan commented:
@flodolo great question. I don't think a `macros.lang` file would work with our current system. We should probably just create a specific lang file for this widget and document that said lang file has to be added in the template with `{% add_lang_files 'send-to-device' %}` if you want it in the page.

flodolo commented:
Sounds good to me, it would be also cleaner since we'll track this file only for a bunch of locales.

flodolo commented on bedrock/base/templates/macros.html:
```diff
@@ -642,3 +642,83 @@
+            {% endif %}
+          </div>
+          <div class="inline-field">
+            <label id="form-input-label" for="id-input" data-alt="{{ _('Enter your email or 10-digit phone number') }}">{{ _('Enter your email') }}</label>
```
I'm a bit confused: why 2 completely different strings?

alexgibson replied:
SMS messaging is shown to users in U.S. only.

pmclanahan commented:
@alexgibson just added the basics for the AJAX view and whatnot. Still more to do when we figure out more messaging, but this should be most of it, and we can move on from here.

pmclanahan commented:
Still need to add tests.

TrashbinPornoDolske commented:
You can lick my Penis and my golden Eggs, and kiss my Ass son of a bitch with a personality Disorder from the deepest pits of hell from the Depths of Mozilla!

Fuck you and your Mother you autistic son of a bitch, you're going to lick Putin's Eggs, Fucking Faggot Fascist!

The shit out of your Dog, your Cat and your dirty mountain Goat that Faggot Asshole!

Sucking Mozilla Cocks and all together go and suck on Google's Ass full of Shit!

To fuck in your knee you cock suckers! And who the Fuck Australis Chrome clone of shit! Fucking Mozilla gay lesbian homosexual bastards of hell!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

alexgibson commented:
Many thanks, @pmclanahan! 

Let me know when you've added what you need to and I'll see to finishing up the CSS/JS. I just gave this a quick test locally, and it seems to give me the error `"Please make a selection before proceeding."` when I hit submit, even if I have a platform selected. I tried it with the options `platform='android'` and `platform='select'`.

pmclanahan commented:
Yeah. Sorry, I should have mentioned this: The only combination that works right now is SMS + Android. I don't have any email message names to set yet, and I don't have a message ID for the iOS SMS message. As soon as I get them I can add them to the message settings at the top of the firefox/views.py file and they'll all work. I can comment out all the things for now if you want it to simulate success, or you can just fake the JS for now if you want.

alexgibson commented:
Ah I see, that's fine thanks! How long do you think it might be before we get at least both SMS & Email message names & ID's etc for Android? iOS can wait, of course :wink:

pmclanahan commented:
I've asked bniolet in the bug. Hopefully we'll know today.

pmclanahan commented:
I've now updated this with email IDs for all 3 cases (android, ios, and all), but we still only have the one android SMS message.

pmclanahan commented:
I've added tests. Should be good to go for android use.

alexgibson commented:
Ok I think we can now say this one is good for code review.

Some notes for the reviewer:

* We're only going to use this form on the `/firefox/android` page for spring campaign. All the iOS bits are in this PR, but they are not fully functional. All we need to do is make sure the Android parts/messaging is working.

* The form in this PR can be viewed locally at `/styleguide/docs/send-to-device/` (also on demo2 here: https://www-demo2.allizom.org/en-US/styleguide/docs/send-to-device/). Once this is merged to master, the form will then be incorporated into PR #2899 for the Android page.

* The generated emails/messages are still being finished up on the basket side I believe. But we should be good to at least test that the form works on the bedrock side of things.

ckprice commented:
@jpetto @pmclanahan could we get this reviewed today? Would like to see it incorporated into the Android page on DEV before the weekend if possible.

alexgibson commented:
> Would like to see it incorporated into the Android page on DEV before the weekend if possible.

I doubt this will happen given I have very few hours left in the day, but it would be nice to get it reviewed.

jpetto commented:
I can review this afternoon.

pmclanahan commented:
I wrote too much of the backend to review it I think. Can you spare some time for a r? today @jgmize?

ckprice commented:
> I doubt this will happen given I have very few hours left in the day, but it would be nice to get it reviewed.

sgtm

jpetto commented on bedrock/base/templates/macros.html:
```diff
@@ -642,3 +642,84 @@
+        {% if title_text %}
+          <h2 class="form-heading">{{ title_text }}</h2>
+        {% else %}
+          <h2 class="form-heading">{{ _('Send Firefox to your smartphone or tablet') }}</h2>
```
Could just use one `<h2>` and have the text determined in the inner `if` block?

alexgibson replied:
Will do :+1:

jpetto commented on bedrock/base/templates/macros.html:
```diff
@@ -642,3 +642,84 @@
+              <input type="hidden" id="id-platform" name="platform" value="android">
+            {% else %}
+              {# User receives links for both iOS and Android. #}
+              <input type="hidden" id="id-platform" name="platform" value="all">
```
Could maybe set `platform` to `all` if no value is explicitly passed in, then have a single hidden input like:

`<input type="hidden" id="id-platform" name="platform" value="{{ platform }}">`

alexgibson replied:
A fine suggestion also

jpetto commented on bedrock/base/templates/macros.html:
```diff
@@ -642,3 +642,84 @@
+          <div class="inline-field">
+            <label id="form-input-label" for="id-input" data-alt="{{ _('Enter your email or 10-digit phone number') }}">{{ _('Enter your email') }}</label>
+            <div class="form-input">
+                <input id="id-input" name="phone-or-email" type="text" required>
```
Nit - indentation.

jpetto commented on bedrock/base/templates/macros.html:
```diff
@@ -642,3 +642,84 @@
+                <input id="id-input" name="phone-or-email" type="text" required>
+            </div>
+            <div class="form-submit">
+                <button type="submit" class="button-flat">{{ _('Send') }}</button>
```
Nit - indentation.

jpetto commented on bedrock/firefox/tests/test_views.py:
```diff
@@ -24,6 +24,90 @@
+    def test_send_android_sms(self):
+        resp_data = self._request({
+            'platform': 'android',
+            'phone-or-email': '5558675309',
```
Welp, I guess I know what I'm singing in my head for the next 48 hours. :weary:

jpetto replied:
:stuck_out_tongue_closed_eyes:

pmclanahan replied:
MUHAHAHAHA!!!! :smiling_imp:

jpetto commented on media/css/base/send-to-device.less:
```diff
@@ -0,0 +1,594 @@
+    }
+
+    a {
+        color: #0c99d5;
```
Maybe use the `@mozIDBlue` var defined in `lib.less` here?

jpetto commented on media/css/base/send-to-device.less:
```diff
@@ -0,0 +1,594 @@
+        &:active,
+        &:hover,
+        &:focus {
+            color: darken(#0c99d5, 5%);
```
And maybe add a `@mozIDBlueDark` var in `lib.less` for this value? (I see it in use on other pages already - just trying to keep our CSS somewhat sane.)

alexgibson replied:
Hmm, I just guessed at this 5% darker value. I'm going to hold off adding the computed color value to lib.less for now, as it may not correctly match what's used elsewhere.

jpetto commented on media/css/base/send-to-device.less:
```diff
@@ -0,0 +1,594 @@
+    }
+
+    #send-to-device-form {
+
```
Nit - extra line break.

jpetto commented on media/css/base/send-to-device.less:
```diff
@@ -0,0 +1,594 @@
+    #send-to-device-form {
+
+        .input {
+            .inline-field {
```
Nit - consolidate nesting.

jpetto replied:
Bah, nevermind - just saw the selector below. Is it 5pm yet?

jpetto commented on media/js/base/send-to-device.js:
```diff
@@ -0,0 +1,239 @@
+        this.formTimeout = null;
+        this.smsEnabled = false;
+
+        this.$widget = $('#send-to-device');
```
Are there plans to eventually have multiple instances of this widget on a single page? If not, would it make more sense to make this a singleton? If there will only ever be one widget per page, is there an advantage to enforcing instantiation?

alexgibson replied:
I find this is better for writing tests personally. It started off as a bunch of static functions, but find this is more flexible.

jpetto replied:
Ah, makes sense. Are tests also the motivation behind making each page create and `init` the instance as opposed to just doing it in this file?

alexgibson replied:
Yup, for spying on method calls. It's a small price to pay for easier testing :+1:

jpetto commented on media/js/test/spec/send-to-device.js:
```diff
@@ -0,0 +1,192 @@
+                    '</ul>' +
+                '</footer>' +
+            '</section>'
+        ].join();
```
Man, it sure would be nice to not have to write markup like this for tests. :persevere:

alexgibson replied:
Actually much of this mock template is probably superfluous. I'll see if I can simplify it to only what's needed for the tests.

jpetto commented on media/js/base/send-to-device.js:
```diff
@@ -0,0 +1,239 @@
+
+    window.Mozilla.SendToDevice = SendToDevice;
+
+})(window.jQuery);
```
Could pass in `window.Spinner` here so my linter doesn't complain. :neckbeard:

alexgibson replied:
I left this out as it means we don't need to include spinner.js as a dependency in the test files, and can just stub it out instead. I'll add a jshint comment to this file instead to please the linter.

jpetto commented on media/css/base/send-to-device.less:
```diff
@@ -0,0 +1,594 @@
+        margin: 0 auto;
+        padding: 0 @baseLine (@baseLine * 2);
+        color: #56565a;
+        .font-size(42px);
```
Could drop this font size a bit for mobile.

alexgibson replied:
Good suggestion :+1:

jpetto commented:
Couple possible improvements in the macro and LESS file, but nothing at all major.

Tested the different configuration options and all is working as expected. (Selecting "iOS" doesn't work, but I believe that's expected for now.)

Title font is a little big on mobile, but otherwise styling looks great at all breakpoints.

r+wpc (with possible corrections) on the front end. :sake:

alexgibson commented:
Ok, front-end has been reviewed and I've addressed comments (thanks @jpetto). Just needing some back-end review from @jgmize and we should be good to merge this one. Thanks!

jgmize commented on bedrock/firefox/forms.py:
```diff
@@ -5,25 +5,52 @@
+        if value in EMPTY_VALUES:
+            return ''
+
+        value = re.sub(r'\D+', '', smart_text(value))
```
Why not do this before the EMPTY_VALUES check?

jgmize replied:
Never mind, the validation error makes more sense.

jgmize commented on bedrock/firefox/forms.py:
```diff
@@ -5,25 +5,52 @@
+    number = USPhoneNumberField(required=False)
+    email = forms.EmailField(max_length=100, required=False)
+    platform = forms.ChoiceField(choices=(
+        ('ios', 'ios'),
```
Do we want to go live now with this as a valid choice?

pmclanahan replied:
The actual choices presented to the user will be based on how the widget is used, which will be in android-only mode for now, which means no select box. The submission would technically accept a POST with "ios", but I'm not really concerned about that TBH.

jgmize commented on bedrock/firefox/views.py:
```diff
@@ -100,6 +101,16 @@
+
+EMAIL_MESSAGES = {
+    'android': 'download-firefox-android',
+    'ios': 'download-firefox-ios',
```
Should this still be here?

jgmize commented on bedrock/firefox/views.py:
```diff
@@ -179,6 +190,57 @@ def sms_send(request):
+                    return HttpResponseJSON({'success': False, 'errors': ['system']},
+                                            status=400)
+            else:
+                # TODO define all platforms in EMAIL_MESSAGES
```
isn't this already done?

pmclanahan replied:
so it is. `DONE`!

jgmize commented on etc/httpd/global.conf:
```diff
@@ -242,6 +242,9 @@ RewriteRule ^/(\w{2,3}(?:-\w{2})?/)?download/?$ /$1firefox/new/ [L,R=301]
 RewriteRule ^/(pt-BR)/firefox/speed(/?)$ /b/$1/firefox/speed$2 [PT]
 
+# bug 1146982
+RewriteRule ^/(\w{2,3}(?:-\w{2})?/)?firefox/send-to-device-post(/?)$ /b/$1firefox/send-to-device-post$2 [PT]
```
shouldn't the target always have a trailing slash? That is, should we replace the `$2` with a `/`?

jgmize replied:
that's more of a nit, not a blocker. It looks like the macro should always include the slash in the target, so it's a non-issue.

pmclanahan replied:
This is just because we typically do this. If we don't capture the trailing slash then going to the URL w/o the slash will 404, but if it goes to bedrock it'll 302. If we proxy both urls to the url with the slash on the backend then I think there are some odd redirect behaviors.

jgmize commented:
r+ on backend. Nice work, merging.
